pkg/skaffold/deploy/cloudrun/accessor.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+	"github.com/GoogleContainerTools/skaffold/proto/v1"
+)
+
+var (
+	retrieveAvailablePort = util.GetAvailablePort
+	gcloudInstalled       = findGcloud
+)
+
+type resourceTracker struct {
+	resources      []forwardedResource
+	forwardedPorts *util.PortSet
+}
+
+type forwardedResource struct {
+	name    RunResourceName
+	cancel  context.CancelFunc
+	started bool
+	port    int
+}
+
+type forwarder interface {
+	// Start initiates the forwarder's operation. It should not return until any ports have been allocated.
+	Start(ctx context.Context, out io.Writer) error
+	Stop()
+}
+
+// RunAccessor is an access.Accessor for Cloud Run resources
+// It uses `gcloud run proxy``to enable port forwarding for Cloud Run. This makes it easier to call IAM-protected Cloud Run services
+// by going through localhost. In order to set up forwarding, the services must have their ingress setting set to "all", gcloud  must be
+// installed and on the path, and the currently configured gcloud user has run.services.invoke permission on the services being proxied
+type RunAccessor struct {
+	resources  *resourceTracker
+	forwarders []forwarder
+	singleRun  singleflight.Group
+	label      string
+}
+
+// NewAccessor creates a new RunAccessor to port forward Cloud Run services
+func NewAccessor(cfg Config, label string) *RunAccessor {
+	var forwarders []forwarder
+	resources := &resourceTracker{forwardedPorts: &util.PortSet{}}
+	options := cfg.PortForwardOptions()
+	if options.ForwardServices(cfg.Mode()) {
+		forwarders = append(forwarders, &runProxyForwarder{resources: resources})
+	}
+	return &RunAccessor{
+		resources: resources, forwarders: forwarders, label: label, singleRun: singleflight.Group{}}
+}
+
+// AddResource tracks an additional resource to port forward
+func (r *RunAccessor) AddResource(resource RunResourceName) {
+	r.resources.resources = append(r.resources.resources, forwardedResource{name: resource, started: false})
+}
+
+// Start begins port forwarding for the tracked Cloud Run services.
+func (r *RunAccessor) Start(ctx context.Context, out io.Writer) error {
+	if r == nil {
+		return nil
+	}
+	_, err, _ := r.singleRun.Do(r.label, func() (interface{}, error) {
+		return struct{}{}, r.start(ctx, out)
+	})
+	return err
+}
+
+func (r *RunAccessor) start(ctx context.Context, out io.Writer) error {
+	for _, forwarder := range r.forwarders {
+		if err := forwarder.Start(ctx, out); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Stop terminates port forwarding for all tracked Cloud Run resources.
+func (r *RunAccessor) Stop() {
+	for _, forwarder := range r.forwarders {
+		forwarder.Stop()
+	}
+}
+
+func findGcloud() bool {
+	cmd := exec.Command("gcloud", "--version")
+	return cmd.Run() == nil
+}
+
+type runProxyForwarder struct {
+	resources *resourceTracker
+}
+
+func (r *runProxyForwarder) Start(ctx context.Context, out io.Writer) error {
+	if !gcloudInstalled() {
```

Comment:
likely makes sense to also check that the cloud run proxy is installed here and add a related actionable error as well

Reply:
Hmm, this is a bit complex. If gcloud is installed manually, it seems it'll automatically install it. It only runs into a problem if you need root to install components or it was installed through a package manager. I can add the check, but it'll lead to a worse experience in one case.

Reply:
I think it is ok to leave as is, the context was mainly around wanting to make sure that this error is actionable as only actionable errors are properly bubbled up in our metrics pipeline s.t. Cloud Code can piece out how many users might hit this.

I'm not sure I understand how doing an explicit check leads to a worse experience in one case vs the other atm though